File: src/testGen.cpp
#include <fstream>
#include <iostream>
#include <random>
#include <string>
#include <vector>
#include <iomanip>

#include "cracker.h"
#include "rules.h"
#include "sha256.h"
#include "util.h"

std::array<std::string, 4> salts = {"agr", "beg", "chr", "drg"};

int main() {
    std::string hashes_filename = "doc/hashes.txt";
    std::string wordlist_filename = "doc/wordlist.txt";
    std::string rules_filename = "doc/rules.txt";

    std::ifstream wordlist_file(wordlist_filename);
    std::ifstream rules_file(rules_filename);
    std::ofstream hashes_file(hashes_filename);

    std::vector<std::string> words;
    std::vector<std::string> rules;
    std::string word;
    std::string rule;

    while (std::getline(wordlist_file, word)) words.push_back(word);
    while (std::getline(rules_file, rule)) rules.push_back(rule);
    
    for (int i = 0; i < 100; i++) {
        std::string salt = salts[i%4];
        std::string complete_string = words[i] + salt;
        complete_string.resize(100, '\0');
        // std::cout << complete_string << '\n';
        char candidate[100];
        char hex[100], tmp[100];
        rules_apply(complete_string.data(), rules[i%7].data(), candidate,
                        complete_string.size());
        SHA256 ctx;
        sha256(&ctx, (BYTE*)(candidate), strlen(candidate));
        
        memcpy(tmp, ctx.b, 32);
        tmp[32] = '\0';
        char show[100];
        utf8ToHex(tmp, show, 32);

        for (int i = 0; show[i]; i++)
            if(isalpha(show[i]))
                show[i] = tolower((unsigned char)show[i]);

        hashes_file << show << ":" << salt << '\n';
    }
}

File: testGen.py
import hashlib

salts = ["agr", "beg", "chr", "drg"]

def utf8_to_hex(input_str):
    # 將 UTF-8 字串轉換為十六進位字串
    return ''.join(f'{ord(char):02X}' for char in input_str)

def hex_to_ascii(hex_str):
    # 將每 4 個字節轉換為對應的 ASCII 字元
    if len(hex_str) % 4 != 0:
        raise ValueError("Hex string length must be a multiple of 4.")
    ascii_str = ''.join(chr(int(hex_str[i:i+4], 16)) for i in range(0, len(hex_str), 4))
    return ascii_str

def sha256_hash(input_data):
    # SHA-256 雜湊運算
    sha = hashlib.sha256()
    sha.update(input_data)
    return sha.digest()

def main():
    hashes_filename = "doc/hashes.txt"
    wordlist_filename = "doc/wordlist.txt"
    rules_filename = "doc/rules.txt"

    with open(wordlist_filename, "r") as wordlist_file:
        words = [line.strip() for line in wordlist_file]
    
    with open(rules_filename, "r") as rules_file:
        rules = [line.strip() for line in rules_file]

    for i in range(3):
        salt = salts[i % 4]
        complete_string = words[i] + salt
        hex_str = utf8_to_hex(complete_string)
        for _ in range(100): 
            hex_str = sha256_hash(hex_str.encode()).hex()
        show = hex_to_ascii(hex_str)
        print(f"{show}:{salt}")

if __name__ == "__main__":
    main()
